### PR TITLE
Try unify training and inference

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136888

Previously inference -> inference IR was going through a seperate flow from train -> inference decomposition. This diff unifies them so that we always retrace when decomposing. Joint IR decomp is still going through old flow (inference -> inference) but seems ok for now since it is still in experimental stage. 

Differential Revision: [D63062521](https://our.internmc.facebook.com/intern/diff/D63062521/)

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63062521).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63062521).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63062521).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63062521).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63062521).

File: torch/export/_trace.py [hunk 1]
```diff
@@ -745,7 +745,7 @@ def _maybe_fixup_gm_and_output_node_meta(old_gm, new_gm):
         for node in _mod.graph.nodes:
             if node.op in ["placeholder", "output"]:
                 node.meta.pop("nn_module_stack", None)
-                node.meta.pop("stack_trace", None)
+                node.meta.pop("stack_trace", None)
```
Comment (avikchaudhuri) [inline torch/export/_trace.py:748]: undo

File: torch/export/exported_program.py [hunk 2]
```diff
@@ -355,7 +359,8 @@ def _decompose_and_get_gm_with_new_signature_constants(
                 fake_args.append(node.meta["val"])
 
         fake_args_unwrapped = pytree.tree_unflatten(fake_args, mod._in_spec)
-        fake_mode = _detect_fake_mode_from_gm(mod)
+        # TODO T204030333
```
Comment (avikchaudhuri) [inline torch/export/exported_program.py:362]: Is it because of tracing with fake tensors or something else?

**Review by avikchaudhuri — COMMENTED** [review]
tests are failing, looks like naming placeholders is off and maybe other changes to ops too?

File: torch/export/exported_program.py (hunk 2 above)
Comment (tugsbayasgalan) [reply torch/export/exported_program.py:362]: I didn't investigate further. Possibly..

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63062521).

**Review by avikchaudhuri — APPROVED** [review]

**Pull request closed without merging**